<div style="font-size:20px; color:pink; margin:20px">
全体の流れを意識して機械学習に取り組む
</div>
<div style="font-size:14px; color:white; margin:20px">
機械学習を直感的に理解してもらうために、簡易的なデータを準備し、機械学習の部分だけを取り扱ってきました。<br>
しかし、本来は、<br>
 1)機械学習を使うための目的や解決すべき課題の設計<br>
 2)データ設計<br>
 3)データの収集・加工<br>
 4)データの把握<br>
 5)機械学習を実行<br>
できます。<br>
今回は、これまでよりも複雑なデータを用意したので、<br>
全体の流れ(1〜5の工程)を意識しながら機械学習に取り組んでいきましょう。<br>
</div>

<div style="font-size:20px; color:pink; margin:20px">
データの準備・概要把握
</div>
<div style="font-size:14px; color:white; margin:20px">
今回は、消費者物価指数から、消費者支出が増加するかを予測(分類)してみます。<br>
データは、これまでと同様に政府統計の総合窓口e-statのデータを使用します。<br>
<br>
目的変数として、月次の消費者支出を家計調査から取得し、<br>
先月よりも支出が増加した場合はフラグ1を、減少した場合はフラグ0を付与しました。<br>
説明変数としては、月次の消費者物価指数を2015年基準消費者物価指数から取得し、加工してあります。<br>
今回は冒頭にあまり細かくデータの説明を行いません。どんなデータなのかを把握していく作業も一緒にやっていきましょう。<br>
</div>

In [2]:
import pandas as pd
data = pd.read_csv('ConsumerExpenditures.csv')
data.head()

Date     消費支出  消費支出増加フラグ  食料-1ヶ月前  住居-1ヶ月前  光熱・水道-1ヶ月前  家具・家事用品-1ヶ月前  \
0   1985年7月  281,604          1     81.1     75.6        87.9         155.5   
1   1985年8月  266,596          0     81.8     75.7        87.8         155.5   
2   1985年9月  249,416          0     81.8     75.8        87.8         155.3   
3  1985年10月  264,096          1     81.3     75.9        87.8         155.7   
4  1985年11月  257,024          0     82.9     76.0        87.8         155.8   

   被服及び履物-1ヶ月前  保健医療-1ヶ月前  交通・通信-1ヶ月前    ...     食料-6ヶ月前  住居-6ヶ月前  光熱・水道-6ヶ月前  \
0         82.0       80.3        99.1    ...        81.2     74.6        87.8   
1         81.3       80.4        99.0    ...        80.9     74.7        87.6   
2         78.2       80.4        98.8    ...        81.2     74.9        87.6   
3         81.7       80.4        98.8    ...        81.5     75.1        87.8   
4         82.7       80.5        99.2    ...        81.2     75.5        87.8   

   家具・家事用品-6ヶ月前  被服及び履物-6ヶ月前  保健医療-6ヶ月前  交通・通信-6ヶ月前  教育-6ヶ月前  教養娯楽-6ヶ月前  \
0         155.7         79.4       78.8        97.9     61.1      101.7   
1         155.0         78.0       78.8        97.5     61.1      100.8   
2         155.2         79.3       80.1        97.9     61.1      101.1   
3         155.5         80.7       80.2        97.9     63.7      101.9   
4         155.5         82.0       80.3        99.3     63.7      101.4   

   諸雑費-6ヶ月前  
0      74.4  
1      74.5  
2      74.5  
3      74.5  
4      74.6  

[5 rows x 63 columns]

In [3]:
data.count()

Date            387
消費支出            387
消費支出増加フラグ       387
食料-1ヶ月前         387
住居-1ヶ月前         387
光熱・水道-1ヶ月前      387
家具・家事用品-1ヶ月前    387
被服及び履物-1ヶ月前     387
保健医療-1ヶ月前       387
交通・通信-1ヶ月前      387
教育-1ヶ月前         387
教養娯楽-1ヶ月前       387
諸雑費-1ヶ月前        387
食料-2ヶ月前         387
住居-2ヶ月前         387
光熱・水道-2ヶ月前      387
家具・家事用品-2ヶ月前    387
被服及び履物-2ヶ月前     387
保健医療-2ヶ月前       387
交通・通信-2ヶ月前      387
教育-2ヶ月前         387
教養娯楽-2ヶ月前       387
諸雑費-2ヶ月前        387
食料-3ヶ月前         387
住居-3ヶ月前         387
光熱・水道-3ヶ月前      387
家具・家事用品-3ヶ月前    387
被服及び履物-3ヶ月前     387
保健医療-3ヶ月前       387
交通・通信-3ヶ月前      387
               ... 
食料-4ヶ月前         387
住居-4ヶ月前         387
光熱・水道-4ヶ月前      387
家具・家事用品-4ヶ月前    387
被服及び履物-4ヶ月前     387
保健医療-4ヶ月前       387
交通・通信-4ヶ月前      387
教育-4ヶ月前         387
教養娯楽-4ヶ月前       387
諸雑費-4ヶ月前        387
食料-5ヶ月前         387
住居-5ヶ月前         387
光熱・水道-5ヶ月前      387
家具・家事用品-5ヶ月前    387
被服及び履物-5ヶ月前     387
保健医療-5ヶ月前       387
交通・通信-5ヶ月前      387
教育-5ヶ月前         387
教養娯楽-5ヶ月前       387


<div style="font-size:14px; color:white; margin:20px">
countを指定することで、列名とその列に入っているデータ数が表示されます。<br>
どのデータも387とあるように、今回のデータは387行存在することがわかります。<br>
これで、行数に関しても、これまでの50行程度から大きく増えていることが理解できたかと思います。<br>
また、今回のデータでは存在しないのですが、データが入力されておらずデータ欠損(null)が生じていることが多々あります。<br>
countを実行すると、データの欠損値は数に含まれないので、欠損があるかどうかの把握を簡易的に行えます。<br>
<br>
さて、列名もcountによって大分見えてきましたが、せっかくなので列名だけ取り出してみましょう。<br>

In [4]:
data.columns

Index(['Date', '消費支出', '消費支出増加フラグ', '食料-1ヶ月前', '住居-1ヶ月前', '光熱・水道-1ヶ月前',
       '家具・家事用品-1ヶ月前', '被服及び履物-1ヶ月前', '保健医療-1ヶ月前', '交通・通信-1ヶ月前', '教育-1ヶ月前',
       '教養娯楽-1ヶ月前', '諸雑費-1ヶ月前', '食料-2ヶ月前', '住居-2ヶ月前', '光熱・水道-2ヶ月前',
       '家具・家事用品-2ヶ月前', '被服及び履物-2ヶ月前', '保健医療-2ヶ月前', '交通・通信-2ヶ月前', '教育-2ヶ月前',
       '教養娯楽-2ヶ月前', '諸雑費-2ヶ月前', '食料-3ヶ月前', '住居-3ヶ月前', '光熱・水道-3ヶ月前',
       '家具・家事用品-3ヶ月前', '被服及び履物-3ヶ月前', '保健医療-3ヶ月前', '交通・通信-3ヶ月前', '教育-3ヶ月前',
       '教養娯楽-3ヶ月前', '諸雑費-3ヶ月前', '食料-4ヶ月前', '住居-4ヶ月前', '光熱・水道-4ヶ月前',
       '家具・家事用品-4ヶ月前', '被服及び履物-4ヶ月前', '保健医療-4ヶ月前', '交通・通信-4ヶ月前', '教育-4ヶ月前',
       '教養娯楽-4ヶ月前', '諸雑費-4ヶ月前', '食料-5ヶ月前', '住居-5ヶ月前', '光熱・水道-5ヶ月前',
       '家具・家事用品-5ヶ月前', '被服及び履物-5ヶ月前', '保健医療-5ヶ月前', '交通・通信-5ヶ月前', '教育-5ヶ月前',
       '教養娯楽-5ヶ月前', '諸雑費-5ヶ月前', '食料-6ヶ月前', '住居-6ヶ月前', '光熱・水道-6ヶ月前',
       '家具・家事用品-6ヶ月前', '被服及び履物-6ヶ月前', '保健医療-6ヶ月前', '交通・通信-6ヶ月前', '教育-6ヶ月前',
       '教養娯楽-6ヶ月前', '諸雑費-6ヶ月前'],
      dtype='object')

<div style="font-size:14px; color:white; margin:20px">
これによって、列名がすべて取得できました。<br>
この列名を見ていくと、<br>
今回の目的変数である消費支出増加フラグと、それ以降の説明変数に大きく分かれることがわかります。<br>
説明変数を見ていくと、第3回で扱った消費者物価指数の項目と同じです。<br>
ただし、当月(Date)に対して1カ月前から6カ月前まで遡って消費者物価指数が用意されています。<br>
このようなデータを用意した理由は、冒頭に述べた1)機械学習の目的と関連しています。<br>
<br>
何度か述べていますが、機械学習の本来の目的は、未知な事例を予測することにあります。<br>
今回の例でいうと、来月の消費者支出を予測することです。<br>
来月の予測を行う際に、来月のデータを使うことはできるのでしょうか。<br>
もうおわかりのように、来月のデータはまだ存在しないため、今月のデータのみで、来月の予測を行う必要があります。<br>
そのため、過去に遡って消費者物価指数を用意しているのです。<br>
物価が下がれば、お金を出しやすくなるというのはイメージしやすいかと思います。<br>
今回は取り扱いませんが、<br>
正確には、過去の消費者物価指数が支出に寄与しているのをグラフ化したりして確認するとなお良いと思います。<br>
何カ月前の物価が支出に寄与しているかわからないため、<br>
1〜6カ月前までのデータをそれぞれ抽出し機械学習できる形に<br>
データを設計し(2.データ設計工程)、<br>
加工を行なっています(3.データの収集・加工工程)。<br>
</div>

<div style="font-size:20px; color:pink; margin:20px">
データの準備
</div>
<div style="font-size:14px; color:white; margin:20px">
それでは、機械学習のために、データ準備をしていきましょう。
先ほど調べてわかったかと思いますが、
目的変数は消費支出増加フラグ、説明変数は食料から諸雑費、1カ月前から6カ月前までのデータとなります
</div>

In [11]:
# 読み込んだデータを残すために、data_tmpを作成
data_tmp = data.copy()
# 不要なデータを削除
del data_tmp['Date']
del data_tmp['消費支出']
# 説明変数data_tmp_X、目的変数data_tmp_Yを作成
data_tmp_X = data_tmp.copy()
del data_tmp_X['消費支出増加フラグ']
data_tmp_Y = data_tmp['消費支出増加フラグ']
# datatmp_X.columnsで、列がしっかり用意できているか確認
data_tmp_X.columns

Index(['食料-1ヶ月前', '住居-1ヶ月前', '光熱・水道-1ヶ月前', '家具・家事用品-1ヶ月前', '被服及び履物-1ヶ月前',
       '保健医療-1ヶ月前', '交通・通信-1ヶ月前', '教育-1ヶ月前', '教養娯楽-1ヶ月前', '諸雑費-1ヶ月前',
       '食料-2ヶ月前', '住居-2ヶ月前', '光熱・水道-2ヶ月前', '家具・家事用品-2ヶ月前', '被服及び履物-2ヶ月前',
       '保健医療-2ヶ月前', '交通・通信-2ヶ月前', '教育-2ヶ月前', '教養娯楽-2ヶ月前', '諸雑費-2ヶ月前',
       '食料-3ヶ月前', '住居-3ヶ月前', '光熱・水道-3ヶ月前', '家具・家事用品-3ヶ月前', '被服及び履物-3ヶ月前',
       '保健医療-3ヶ月前', '交通・通信-3ヶ月前', '教育-3ヶ月前', '教養娯楽-3ヶ月前', '諸雑費-3ヶ月前',
       '食料-4ヶ月前', '住居-4ヶ月前', '光熱・水道-4ヶ月前', '家具・家事用品-4ヶ月前', '被服及び履物-4ヶ月前',
       '保健医療-4ヶ月前', '交通・通信-4ヶ月前', '教育-4ヶ月前', '教養娯楽-4ヶ月前', '諸雑費-4ヶ月前',
       '食料-5ヶ月前', '住居-5ヶ月前', '光熱・水道-5ヶ月前', '家具・家事用品-5ヶ月前', '被服及び履物-5ヶ月前',
       '保健医療-5ヶ月前', '交通・通信-5ヶ月前', '教育-5ヶ月前', '教養娯楽-5ヶ月前', '諸雑費-5ヶ月前',
       '食料-6ヶ月前', '住居-6ヶ月前', '光熱・水道-6ヶ月前', '家具・家事用品-6ヶ月前', '被服及び履物-6ヶ月前',
       '保健医療-6ヶ月前', '交通・通信-6ヶ月前', '教育-6ヶ月前', '教養娯楽-6ヶ月前', '諸雑費-6ヶ月前'],
      dtype='object')

In [12]:
# データを訓練データと、テストデータに分割
# 分割には、traintestsplitを使用
from sklearn.model_selection import train_test_split
# サイズは、訓練70%、テスト30%に分割
X_train, X_test, Y_train, Y_test = train_test_split(data_tmp_X, data_tmp_Y, random_state=0, test_size=0.3)
# 分割できているか確かめるために、lenを用いてデータ数を単純に合算し、printで出力
print(len(X_train))
print(len(Y_train))
print(len(X_test))
print(len(Y_test))

270
270
117
117


<div style="font-size:20px; color:pink; margin:20px">
決定木
</div>
<div style="font-size:14px; color:white; margin:20px">
まずは、決定木による予測モデル作成をやっていきましょう。
</div>

In [13]:
# 決定木のインポート
from sklearn.tree import DecisionTreeClassifier
# 予測モデルの定義(今回の木の深さは、とりあえず3)
treeModel = DecisionTreeClassifier(max_depth=3, random_state=0)
# 予測モデルを作成
treeModel.fit(X_train, Y_train)
# 精度を把握するために、accuracy_scoreで最も簡易的なモデル評価(精度)を行った
from sklearn import metrics
print(metrics.accuracy_score(Y_train, treeModel.predict(X_train)))
print(metrics.accuracy_score(Y_test, treeModel.predict(X_test)))

0.785185185185
0.606837606838


<div style="font-size:14px; color:white; margin:20px">
訓練データで78%、テストデータで60%となりました。<br>
訓練データでの精度が高いため、訓練データに過剰適合(過学習)傾向にありますが、<br>
訓練、テストデータどちらを用いても60%以上の精度は出ています。<br>
<br>
過学習傾向なため、モデルの複雑さを減らしてみましょう。<br>
モデルをシンプルにする方法として、max_depthを小さくする方法があります
</div>

In [14]:
treeModel = DecisionTreeClassifier(max_depth=2, random_state=0)
treeModel.fit(X_train, Y_train)
print(metrics.accuracy_score(Y_train, treeModel.predict(X_train)))
print(metrics.accuracy_score(Y_test, treeModel.predict(X_test)))

0.7
0.632478632479


<div style="font-size:14px; color:white; margin:20px">
max_depthを2にすると、訓練データの精度は減少し、テストデータの精度は向上しました。<br>
これによって、精度は下がりましたが、モデルは汎化され、より良いモデルになったと考えられます。<br>
最後に、モデルに寄与している変数を抽出してみましょう。<br>
</div>

In [15]:
Importance = pd.DataFrame({'変数名':data_tmp_X.columns, '重要度':treeModel.feature_importances_})
Importance[Importance['重要度']!=0]

変数名       重要度
24  被服及び履物-3ヶ月前  0.216956
34  被服及び履物-4ヶ月前  0.505538
54  被服及び履物-6ヶ月前  0.277506

<div style="font-size:14px; color:white; margin:20px">
これだけ見ると、すべて衣類関係の物価が寄与しているという結果になりました。<br>
解釈は難しいところですが、衣類関連だけですべての消費支出を決めているとは考えにくいです。<br>
まだモデル精度が低いことも考慮する必要があると思います。<br>
しかし、このデータセットを使用する限り、これ以上決定木のモデル精度が向上させるのは困難なため、<br>
ロジスティック回帰に挑戦し、モデル精度を比較してみましょう。<br>
</div>

<div style="font-size:20px; color:pink; margin:20px">
ロジスティック回帰
</div>
<div style="font-size:14px; color:white; margin:20px">
使用するデータは、決定木と全く同じで良いので、訓練データ、テストデータに分割されているものを使用
</div>

In [16]:
# ロジスティック回帰のインポート
from sklearn.linear_model import LogisticRegression
# 予測モデルの定義(決定木と異なり、特にパラメータ指定はせず、デフォルトのものを使用)
logModel = LogisticRegression()
# Fitで予測モデルを作成
logModel.fit(X_train, Y_train)
# 精度評価(簡易的)
print(metrics.accuracy_score(Y_train, logModel.predict(X_train)))
print(metrics.accuracy_score(Y_test, logModel.predict(X_test)))

0.833333333333
0.794871794872


<div style="font-size:14px; color:white; margin:20px">
訓練データで83%、テストデータでも79%を示しました。<br>
決定木よりも全体的な精度が高いだけでなく、訓練データとテストデータの精度に開きがないため、<br>
汎化されている良いモデルとなっています<br>
<br>
前回述べたように、手法によって分類の方法が異なるため、今回のケースのように大きくモデル精度が異なることがあります。<br>
いくつかの手法を用いて適切に評価を行い、最適なモデルを作成していくことの重要性を理解できたのではないでしょうか。<br>
<br>
せっかくなので、ロジスティック回帰のモデルについて、もう少し触れていきましょう。<br>
前回、ロジスティック回帰は、決定木のように条件分岐によって線を引いていくのではなく、<br>
あくまでも多変数の式を元に分割線を引くものであるとお伝えしました。<br>
ロジスティック回帰は、線形回帰の関数Y=aX1+bX2+zを0から1の範囲に押し込めてしまう関数を使用しています。<br>
そのため、線形回帰と同様にモデルを作成することは線を引くことであり、線を引くことは係数(a、b、z)を決めることと同義です。<br>
<br>
ロジスティック回帰には、決定木のようにfeature_importancesが存在しませんが、<br>
係数によってある程度変数の寄与がわかるようになっています。早速、出力してみましょう。<br>
</div>

In [18]:
Coef = pd.DataFrame({'変数名':data_tmp_X.columns, '係数':logModel.coef_[0]})
Coef.sort_values('係数')

係数           変数名
37 -2.269903       教育-4ヶ月前
47 -1.047428       教育-5ヶ月前
51 -0.998511       住居-6ヶ月前
58 -0.804435     教養娯楽-6ヶ月前
56 -0.734573    交通・通信-6ヶ月前
32 -0.468284    光熱・水道-4ヶ月前
20 -0.416337       食料-3ヶ月前
10 -0.416337       食料-2ヶ月前
36 -0.382380    交通・通信-4ヶ月前
8  -0.372763     教養娯楽-1ヶ月前
30 -0.354499       食料-4ヶ月前
1  -0.250821       住居-1ヶ月前
41 -0.210186       住居-5ヶ月前
45 -0.205891     保健医療-5ヶ月前
35 -0.157501     保健医療-4ヶ月前
18 -0.154217     教養娯楽-2ヶ月前
28 -0.154217     教養娯楽-3ヶ月前
13 -0.139926  家具・家事用品-2ヶ月前
23 -0.139926  家具・家事用品-3ヶ月前
34 -0.133886   被服及び履物-4ヶ月前
6  -0.131997    交通・通信-1ヶ月前
7  -0.100315       教育-1ヶ月前
4  -0.084352   被服及び履物-1ヶ月前
59 -0.074547      諸雑費-6ヶ月前
9  -0.056208      諸雑費-1ヶ月前
33 -0.055393  家具・家事用品-4ヶ月前
44 -0.054628   被服及び履物-5ヶ月前
43 -0.050417  家具・家事用品-5ヶ月前
25 -0.041718     保健医療-3ヶ月前
15 -0.041718     保健医療-2ヶ月前
52 -0.020099    光熱・水道-6ヶ月前
29  0.012744      諸雑費-3ヶ月前
19  0.012744      諸雑費-2ヶ月前
48  0.029198     教養娯楽-5ヶ月前
2   0.029202    光熱・水道-1ヶ月前
3   0.051487  家具・家事用品-1ヶ月前
5   0.117717     保健医療-1ヶ月前
12  0.139468    光熱・水道-2ヶ月前
22  0.139468    光熱・水道-3ヶ月前
0   0.159640       食料-1ヶ月前
39  0.164011      諸雑費-4ヶ月前
14  0.181036   被服及び履物-2ヶ月前
24  0.181036   被服及び履物-3ヶ月前
49  0.201482      諸雑費-5ヶ月前
46  0.242253    交通・通信-5ヶ月前
42  0.261626    光熱・水道-5ヶ月前
55  0.268067     保健医療-6ヶ月前
31  0.285801       住居-4ヶ月前
54  0.298408   被服及び履物-6ヶ月前
26  0.334295    交通・通信-3ヶ月前
16  0.334295    交通・通信-2ヶ月前
50  0.347027       食料-6ヶ月前
40  0.371799       食料-5ヶ月前
53  0.530350  家具・家事用品-6ヶ月前
11  0.630169       住居-2ヶ月前
21  0.630169       住居-3ヶ月前
38  1.093850     教養娯楽-4ヶ月前
17  1.115173       教育-2ヶ月前
27  1.115173       教育-3ヶ月前
57  1.245450       教育-6ヶ月前

<div style="font-size:14px; color:white; margin:20px">
1行目で、今回作成したモデルから、係数を抽出しています。<br>
これらの係数は、Y=aX1+bX2+zのaやbに相当します。<br>
つまり、それぞれの変数に掛かっており、絶対値が大きければ大きいほど寄与は大きく、<br>
正であれば支出増加に、負であれば支出減少に寄与していることになります。<br>
絶対値が大きい部分には、1カ月前の物価指数は存在しておらず、<br>
1カ月前の物価指数の寄与は小さいことが考えられます。<br>
つまり、支出増減には少なくとも2カ月以上前の物価が大きく影響していると予想できます。<br>
ただし、住宅の物価指数が、正にも負にも入っており、景気の波等が関係している可能性があります。<br>
さらにより良いモデルにしていく場合、データの粒度や説明変数の見直しも必要になると思います。<br>
</div>